pink/tasks/com_task.py

```diff
@@ -0,0 +1,156 @@
+        lm_damping: float = 0.0,
+        gain: float = 1.0,
+    ) -> None:
+        r"""Define a new center of mass task.
+
+        Args:
+            cost: Contribution of position errors to the normalized cost, in
+                :math:`[\mathrm{cost}] / [\mathrm{m}]`. If this is a vector,
+                the cost is anisotropic and each coordinate corresponds to an
+                axis of the frame.
+            lm_damping: Levenberg-Marquardt damping (see class attributes). The
+                default value is conservatively low.
+            gain: Task gain :math:`\alpha \in [0, 1]` for additional low-pass
+                filtering. Defaults to 1.0 (no filtering) for dead-beat
+                control.
+        """
+        super().__init__(
+            cost=np.ones(3),  # updated below
+            gain=gain,
+            lm_damping=lm_damping,
+        )
+        self.target_com = None
+        self.set_cost(cost)
+
+    def set_cost(
+        self, cost: Union[float, Sequence[float], np.ndarray]
+    ) -> None:
+        r"""Set a new cost for all 3D CoM position coordinates.
+
+        Args:
+            cost: Contribution of position errors to the normalized cost, in
+                :math:`[\mathrm{cost}] / [\mathrm{m}]`. If this is a vector,
+                the cost is anisotropic and each coordinate corresponds to an
+                axis of the frame.
+        """
+
+        if isinstance(cost, float):
+            assert cost >= 0.0
+        else:  # not isinstance(cost, float)
+            assert all(cost >= 0.0 for cost in cost)
+        if isinstance(self.cost, np.ndarray):
+            self.cost[0:3] = cost
+
+    def set_target(
+        self,
+        target_com: np.ndarray,
+    ) -> None:
+        """Set the target CoM position in the world frame.
+
+        Args:
+            target_com: Target position of the center of mass in the world
+                frame.
+        """
+        self.target_com = target_com.copy()
+
+    def set_target_from_configuration(
+        self, configuration: Configuration
+    ) -> None:
+        """Set the task target pose from a given robot configuration.
+
+        Args:
+            configuration: Robot configuration used to compute the target CoM.
+        """
+        q = configuration.q.copy()
+        desired_com = pin.centerOfMass(
+            configuration.model, configuration.data, q
+        )
+        self.set_target(desired_com)
+
+    def compute_error(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the CoM task error.
+
+        The error is the difference between the target CoM position and the
+        current CoM position, both expressed in the world frame.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            CoM task error :math:`e(q)`.
+        """
+        if self.target_com is None:
+            raise TargetNotSet("no target set for CoM")
+        q = configuration.q.copy()
```

Comment:
Why the extra copy? This value is not used after calling `pin.centerOfMass`, and I expect the function has no side effect on `q`.

Reply:
Will check it :)